### PR TITLE
runner.ts: Break into separate functions

### PR DESCRIPTION
**Pull request by andy-hanson**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [ ] New feature, bugfix, or enhancement
  - [ ] Includes tests
- [ ] Documentation update

#### Overview of change:

Just a refactor. Breaks apart the class into smaller functions.
Converted it to return a Promise, but the implementation doesn't use Promises yet.

### CODE REVIEW

File: src/runner.ts [hunk 1]
```diff
@@ -102,155 +104,161 @@ export interface IRunnerOptions {
...
-export class Runner {
-    private static trimSingleQuotes(str: string) {
-        return str.replace(/^'|'$/g, "");
-    }
-
-    constructor(private options: IRunnerOptions, private outputStream: NodeJS.WritableStream) { }
+export const enum Status {
+    Ok = 0,
+    Error = 1,
+    Two = 2, // What is this?
```
Comment (andy-hanson) [inline src/runner.ts]: What is this?

File: src/runner.ts [hunk 2]
```diff
@@ -102,155 +104,160 @@ export interface IRunnerOptions {
...
+    Ok = 0,
+    FatalError = 1,
+    LintError = 2,
+}
 
-    public run(onComplete: (status: number) => void) {
-        if (this.options.version) {
-            this.outputStream.write(Linter.VERSION + "\n");
-            return onComplete(0);
+export async function run(options: Options, outputStream: NodeJS.WritableStream): Promise<Status> {
```
Comment (JoshuaKGoldberg) [inline src/runner.ts]: While you're in the area, would it make sense to stop using any `console` messages at all? In addition to passing the CLI `options` and an `outputStream` it makes sense to also pass an `errorStream`.

Comment (JoshuaKGoldberg) [inline src/runner.ts]: (#2587 would benefit from this)

File: src/runner.ts [hunk 3]
```diff
@@ -27,10 +27,12 @@ import {
     IConfigurationFile,
 } from "./configuration";
 import { FatalError } from "./error";
+import { LintResult } from "./index";
 import * as Linter from "./linter";
 import { consoleTestResultsHandler, runTests } from "./test";
+import { arrayify, flatMap } from "./utils";
 
-export interface IRunnerOptions {
+export interface Options {
```
Comment (nchen63) [inline src/runner.ts]: breaking. Can you rename back?

File: src/runner.ts [hunk 4]
```diff
@@ -102,155 +104,161 @@ export interface IRunnerOptions {
...
+    Ok = 0,
+    FatalError = 1,
+    LintError = 2,
+}
 
-    public run(onComplete: (status: number) => void) {
-        if (this.options.version) {
-            this.outputStream.write(Linter.VERSION + "\n");
-            return onComplete(0);
+export async function run(options: Options, outputStream: NodeJS.WritableStream, errorStream: NodeJS.WritableStream): Promise<Status> {
```
Comment (nchen63) [inline src/runner.ts]: to avoid a breaking change and waiting for 6.0, can you put a Runner class as a wrapper? We can remove it in the next major rev

Comment (andy-hanson) [inline src/runner.ts]: According to https://github.com/palantir/tslint/issues/2446#issuecomment-290613113, anything not in `index` is not public.

Comment (andy-hanson) [inline src/runner.ts]: @nchen63 bump

**Comment by adidahiya** [comment]
ping here, can you address @nchen63's comments?

**Comment by andy-hanson** [comment]
According to https://github.com/palantir/tslint/issues/2446#issuecomment-290613113, anything not in `index` is not public.

**Comment by adidahiya** [comment]
fair enough; I assumed it was exported through index.

got a few merge conflicts left though

**Comment by adidahiya** [comment]
new conflicts with 2bc8982bfcc5b852c6288b3b231b63024e0e034a :(

**Pull request merged**